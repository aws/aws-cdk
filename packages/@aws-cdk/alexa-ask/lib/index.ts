// Alexa::ASK CloudFormation Resources:
export * from './ask.generated';
export * from './pipeline-actions';
