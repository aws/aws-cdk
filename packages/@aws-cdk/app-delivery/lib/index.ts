export * from './pipeline-deploy-stack-action';
