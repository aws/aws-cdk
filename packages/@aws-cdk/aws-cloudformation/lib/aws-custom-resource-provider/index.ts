// tslint:disable:no-console
import AWS = require('aws-sdk');
import { AwsSdkCall } from '../aws-custom-resource';

/**
 * Flattens a nested object
 *
 * @param object the object to be flattened
 * @returns a flat object with path as keys
 */
function flatten(object: object): { [key: string]: string } {
  return Object.assign(
    {},
    ...function _flatten(child: any, path: string[] = []): any {
      return [].concat(...Object.keys(child)
        .map(key =>
          typeof child[key] === 'object'
            ? _flatten(child[key], path.concat([key]))
            : ({ [path.concat([key]).join('.')]: child[key] })
      ));
    }(object)
  );
}

/**
 * Converts true/false strings to booleans in an object
 */
function fixBooleans(object: object) {
  return JSON.parse(JSON.stringify(object), (_k, v) => v === 'true'
    ? true
    : v === 'false'
      ? false
      : v);
}

export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent, context: AWSLambda.Context) {
  try {
    console.log(JSON.stringify(event));
    console.log('AWS SDK VERSION: ' + (AWS as any).VERSION);

    let physicalResourceId = (event as any).PhysicalResourceId;
    let data: { [key: string]: string } = {};
    const call: AwsSdkCall | undefined = event.ResourceProperties[event.RequestType];

    if (call) {
      const awsService = new (AWS as any)[call.service](call.apiVersion && { apiVersion: call.apiVersion });

      try {
        const response = await awsService[call.action](call.parameters && fixBooleans(call.parameters)).promise();
        data = flatten(response);
      } catch (e) {
        if (!call.catchErrorPattern || !new RegExp(call.catchErrorPattern).test(e.code)) {
          throw e;
        }
      }

      if (call.physicalResourceIdPath) {
        physicalResourceId = data[call.physicalResourceIdPath];
      } else {
        physicalResourceId = call.physicalResourceId!;
      }
    }

    await respond('SUCCESS', 'OK', physicalResourceId, data);
  } catch (e) {
    console.log(e);
    await respond('FAILED', e.message, context.logStreamName, {});
  }

  function respond(responseStatus: string, reason: string, physicalResourceId: string, data: any) {
    const responseBody = JSON.stringify({
      Status: responseStatus,
      Reason: reason,
      PhysicalResourceId: physicalResourceId,
      StackId: event.StackId,
      RequestId: event.RequestId,
      LogicalResourceId: event.LogicalResourceId,
      NoEcho: false,
      Data: data
    });

    console.log('Responding', responseBody);

    const parsedUrl = require('url').parse(event.ResponseURL);
    const requestOptions = {
      hostname: parsedUrl.hostname,
      path: parsedUrl.path,
      method: 'PUT',
      headers: { 'content-type': '', 'content-length': responseBody.length }
    };

    return new Promise((resolve, reject) => {
      try {
        const request = require('https').request(requestOptions, resolve);
        request.on('error', reject);
        request.write(responseBody);
        request.end();
      } catch (e) {
        reject(e);
      }
    });
  }
}
