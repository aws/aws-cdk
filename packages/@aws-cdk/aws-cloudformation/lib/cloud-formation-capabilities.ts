/**
 * Capabilities that affect whether CloudFormation is allowed to change IAM resources
 * @deprecated use `core.CfnCapabilities`
 */
export enum CloudFormationCapabilities {
  /**
   * No IAM Capabilities
   *
   * Pass this capability if you wish to block the creation IAM resources.
   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities
   */
  NONE = '',

  /**
   * Capability to create anonymous IAM resources
   *
   * Pass this capability if you're only creating anonymous resources.
   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities
   */
  ANONYMOUS_IAM = 'CAPABILITY_IAM',

  /**
   * Capability to create named IAM resources.
   *
   * Pass this capability if you're creating IAM resources that have physical
   * names.
   *
   * `CloudFormationCapabilities.NamedIAM` implies `CloudFormationCapabilities.IAM`; you don't have to pass both.
   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities
   */
  NAMED_IAM = 'CAPABILITY_NAMED_IAM',

  /**
   * Capability to run CloudFormation macros
   *
   * Pass this capability if your template includes macros, for example AWS::Include or AWS::Serverless.
   *
   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_CreateStack.html
   */
  AUTO_EXPAND = 'CAPABILITY_AUTO_EXPAND'
}
