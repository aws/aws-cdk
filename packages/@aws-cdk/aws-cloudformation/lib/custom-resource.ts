import * as lambda from '@aws-cdk/aws-lambda';
import * as sns from '@aws-cdk/aws-sns';
import * as core from '@aws-cdk/core';

/**
 * Collection of arbitrary properties
 */
export type Properties = {[key: string]: any};

/**
 * Configuration options for custom resource providers.
 */
export interface CustomResourceProviderConfig {
  /**
   * The ARN of the SNS topic or the AWS Lambda function which implements this
   * provider.
   */
  readonly serviceToken: string;
}

/**
 * Represents a provider for an AWS CloudFormation custom resources.
 * @deprecated use `core.ICustomResourceProvider`
 */
export interface ICustomResourceProvider {
  /**
   * Called when this provider is used by a `CustomResource`.
   * @param scope The resource that uses this provider.
   * @returns provider configuration
   */
  bind(scope: core.Construct): CustomResourceProviderConfig;
}

/**
 * Represents a provider for an AWS CloudFormation custom resources.
 */
export class CustomResourceProvider implements ICustomResourceProvider {
  /**
   * The Lambda provider that implements this custom resource.
   *
   * We recommend using a lambda.SingletonFunction for this.
   */
  public static fromLambda(handler: lambda.IFunction): CustomResourceProvider {
    return new CustomResourceProvider(handler.functionArn);
  }

  /**
   * The SNS Topic for the provider that implements this custom resource.
   */
  public static fromTopic(topic: sns.ITopic): CustomResourceProvider {
    return new CustomResourceProvider(topic.topicArn);
  }

  /**
   * Use AWS Lambda as a provider.
   * @deprecated use `fromLambda`
   */
  public static lambda(handler: lambda.IFunction) { return this.fromLambda(handler); }

  /**
   * Use an SNS topic as the provider.
   * @deprecated use `fromTopic`
   */
  public static topic(topic: sns.ITopic) { return this.fromTopic(topic); }

  /**
   * @param serviceToken the ServiceToken which contains the ARN for this provider.
   */
  private constructor(public readonly serviceToken: string) { }

  public bind(_: core.Construct): CustomResourceProviderConfig {
    return { serviceToken: this.serviceToken };
  }
}

/**
 * Properties to provide a Lambda-backed custom resource
 * @deprecated use `core.CustomResourceProps`
 */
export interface CustomResourceProps {
  /**
   * The provider which implements the custom resource.
   *
   * You can implement a provider by listening to raw AWS CloudFormation events
   * through an SNS topic or an AWS Lambda function or use the CDK's custom
   * [resource provider framework] which makes it easier to implement robust
   * providers.
   *
   * [resource provider framework]: https://docs.aws.amazon.com/cdk/api/latest/docs/custom-resources-readme.html
   *
   * ```ts
   * // use the provider framework from aws-cdk/custom-resources:
   * provider: new custom_resources.Provider({
   *   onEventHandler: myOnEventLambda,
   *   isCompleteHandler: myIsCompleteLambda, // optional
   * });
   * ```
   *
   * ```ts
   * // invoke an AWS Lambda function when a lifecycle event occurs:
   * provider: CustomResourceProvider.fromLambda(myFunction)
   * ```
   *
   * ```ts
   * // publish lifecycle events to an SNS topic:
   * provider: CustomResourceProvider.fromTopic(myTopic)
   * ```
   */
  readonly provider: ICustomResourceProvider;

  /**
   * Properties to pass to the Lambda
   *
   * @default - No properties.
   */
  readonly properties?: Properties;

  /**
   * For custom resources, you can specify AWS::CloudFormation::CustomResource
   * (the default) as the resource type, or you can specify your own resource
   * type name. For example, you can use "Custom::MyCustomResourceTypeName".
   *
   * Custom resource type names must begin with "Custom::" and can include
   * alphanumeric characters and the following characters: _@-. You can specify
   * a custom resource type name up to a maximum length of 60 characters. You
   * cannot change the type during an update.
   *
   * Using your own resource type names helps you quickly differentiate the
   * types of custom resources in your stack. For example, if you had two custom
   * resources that conduct two different ping tests, you could name their type
   * as Custom::PingTester to make them easily identifiable as ping testers
   * (instead of using AWS::CloudFormation::CustomResource).
   *
   * @see
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cfn-customresource.html#aws-cfn-resource-type-name
   *
   * @default - AWS::CloudFormation::CustomResource
   */
  readonly resourceType?: string;

  /**
   * The policy to apply when this resource is removed from the application.
   *
   * @default cdk.RemovalPolicy.Destroy
   */
  readonly removalPolicy?: core.RemovalPolicy;
}

/**
 * Deprecated.
 * @deprecated use `core.CustomResource`
 */
export class CustomResource extends core.CustomResource {
  constructor(scope: core.Construct, id: string, props: CustomResourceProps) {
    super(scope, id, {
      pascalCaseProperties: true,
      properties: props.properties,
      removalPolicy: props.removalPolicy,
      resourceType: props.resourceType,
      serviceToken: core.Lazy.stringValue({ produce: () => props.provider.bind(this).serviceToken }),
    });
  }
}
