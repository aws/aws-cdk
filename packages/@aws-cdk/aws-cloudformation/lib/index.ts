export * from './cloud-formation-capabilities';
export * from './custom-resource';
export * from './nested-stack';

// AWS::CloudFormation CloudFormation Resources:
export * from './cloudformation.generated';
