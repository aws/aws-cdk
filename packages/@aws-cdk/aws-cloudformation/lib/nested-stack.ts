import * as sns from '@aws-cdk/aws-sns';
import * as core from '@aws-cdk/core';

/**
 * Initialization props for the `NestedStack` construct.
 *
 * @experimental
 */
export interface NestedStackProps {
  /**
   * The set value pairs that represent the parameters passed to CloudFormation
   * when this nested stack is created. Each parameter has a name corresponding
   * to a parameter defined in the embedded template and a value representing
   * the value that you want to set for the parameter.
   *
   * The nested stack construct will automatically synthesize parameters in order
   * to bind references from the parent stack(s) into the nested stack.
   *
   * @default - no user-defined parameters are passed to the nested stack
   */
  readonly parameters?: { [key: string]: string };

  /**
   * The length of time that CloudFormation waits for the nested stack to reach
   * the CREATE_COMPLETE state.
   *
   * When CloudFormation detects that the nested stack has reached the
   * CREATE_COMPLETE state, it marks the nested stack resource as
   * CREATE_COMPLETE in the parent stack and resumes creating the parent stack.
   * If the timeout period expires before the nested stack reaches
   * CREATE_COMPLETE, CloudFormation marks the nested stack as failed and rolls
   * back both the nested stack and parent stack.
   *
   * @default - no timeout
   */
  readonly timeout?: core.Duration;

  /**
   * The Simple Notification Service (SNS) topics to publish stack related
   * events.
   *
   * @default - notifications are not sent for this stack.
   */
  readonly notifications?: sns.ITopic[];
}

/**
 * A CloudFormation nested stack.
 *
 * When you apply template changes to update a top-level stack, CloudFormation
 * updates the top-level stack and initiates an update to its nested stacks.
 * CloudFormation updates the resources of modified nested stacks, but does not
 * update the resources of unmodified nested stacks.
 *
 * Furthermore, this stack will not be treated as an independent deployment
 * artifact (won't be listed in "cdk list" or deployable through "cdk deploy"),
 * but rather only synthesized as a template and uploaded as an asset to S3.
 *
 * Cross references of resource attributes between the parent stack and the
 * nested stack will automatically be translated to stack parameters and
 * outputs.
 *
 * @experimental
 */
export class NestedStack extends core.NestedStack {
  constructor(scope: core.Construct, id: string, props: NestedStackProps = { }) {
    super(scope, id, {
      parameters: props.parameters,
      timeout: props.timeout,
      notificationArns: props.notifications ? props.notifications.map(n => n.topicArn) : undefined,
    });
  }
}
