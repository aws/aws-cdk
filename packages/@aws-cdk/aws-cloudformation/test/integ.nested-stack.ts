import * as lambda from '@aws-cdk/aws-lambda';
import * as sns from '@aws-cdk/aws-sns';
import * as sns_subscriptions from '@aws-cdk/aws-sns-subscriptions';
import * as sqs from '@aws-cdk/aws-sqs';
import { App, CfnParameter, Construct, Stack } from '@aws-cdk/core';
import * as cfn from '../lib';

/* eslint-disable cdk/no-core-construct */

interface MyNestedStackProps {
  readonly subscriber?: sqs.Queue;
  readonly siblingTopic?: sns.Topic; // a topic defined in a sibling nested stack
  readonly topicCount: number;
  readonly topicNamePrefix: string;
}

class MyNestedStack extends cfn.NestedStack {
  constructor(scope: Construct, id: string, props: MyNestedStackProps) {
    const topicNamePrefixLogicalId = 'TopicNamePrefix';

    super(scope, id, {
      parameters: {
        [topicNamePrefixLogicalId]: props.topicNamePrefix, // pass in a parameter to the nested stack
      },
    });

    const topicNamePrefixParameter = new CfnParameter(this, 'TopicNamePrefix', { type: 'String' });

    for (let i = 0; i < props.topicCount; ++i) {
      const topic = new sns.Topic(this, `topic-${i}`, { displayName: `${topicNamePrefixParameter.valueAsString}-${i}` });

      // since the subscription resources are defined in the subscriber's stack, this
      // will add an SNS subscription resource to the parent stack that reference this topic.
      if (props.subscriber) {
        topic.addSubscription(new sns_subscriptions.SqsSubscription(props.subscriber));
      }
    }

    if (props.subscriber) {
      new lambda.Function(this, 'fn', {
        runtime: lambda.Runtime.NODEJS_10_X,
        code: lambda.Code.inline('console.error("hi")'),
        handler: 'index.handler',
        environment: {
          TOPIC_ARN: props.siblingTopic ? props.siblingTopic.topicArn : '',
          QUEUE_URL: props.subscriber.queueUrl, // nested stack references a resource in the parent
        },
      });
    }
  }
}

class MyTestStack extends Stack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const queue = new sqs.Queue(this, 'SubscriberQueue');

    new MyNestedStack(this, 'NestedStack1', { topicCount: 3, topicNamePrefix: 'Prefix1', subscriber: queue });
    new MyNestedStack(this, 'NestedStack2', { topicCount: 2, topicNamePrefix: 'Prefix2' });
  }
}

const app = new App();
new MyTestStack(app, 'nested-stacks-test');
app.synth();
