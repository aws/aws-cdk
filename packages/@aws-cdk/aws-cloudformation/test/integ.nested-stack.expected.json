{
  "Resources": {
    "SubscriberQueueC193DC66": {
      "Type": "AWS::SQS::Queue"
    },
    "SubscriberQueuePolicy25A0799E": {
      "Type": "AWS::SQS::QueuePolicy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52",
                      "Outputs.nestedstackstestNestedStack1topic02C2D1254Ref"
                    ]
                  }
                }
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com"
              },
              "Resource": {
                "Fn::GetAtt": [
                  "SubscriberQueueC193DC66",
                  "Arn"
                ]
              }
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52",
                      "Outputs.nestedstackstestNestedStack1topic1474E5389Ref"
                    ]
                  }
                }
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com"
              },
              "Resource": {
                "Fn::GetAtt": [
                  "SubscriberQueueC193DC66",
                  "Arn"
                ]
              }
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52",
                      "Outputs.nestedstackstestNestedStack1topic22C710DC4Ref"
                    ]
                  }
                }
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com"
              },
              "Resource": {
                "Fn::GetAtt": [
                  "SubscriberQueueC193DC66",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "Queues": [
          {
            "Ref": "SubscriberQueueC193DC66"
          }
        ]
      }
    },
    "SubscriberQueuenestedstackstestNestedStack1topic089C5EB1396F65087": {
      "Type": "AWS::SNS::Subscription",
      "Properties": {
        "Protocol": "sqs",
        "TopicArn": {
          "Fn::GetAtt": [
            "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52",
            "Outputs.nestedstackstestNestedStack1topic02C2D1254Ref"
          ]
        },
        "Endpoint": {
          "Fn::GetAtt": [
            "SubscriberQueueC193DC66",
            "Arn"
          ]
        }
      }
    },
    "SubscriberQueuenestedstackstestNestedStack1topic1150E1A929A2C267E": {
      "Type": "AWS::SNS::Subscription",
      "Properties": {
        "Protocol": "sqs",
        "TopicArn": {
          "Fn::GetAtt": [
            "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52",
            "Outputs.nestedstackstestNestedStack1topic1474E5389Ref"
          ]
        },
        "Endpoint": {
          "Fn::GetAtt": [
            "SubscriberQueueC193DC66",
            "Arn"
          ]
        }
      }
    },
    "SubscriberQueuenestedstackstestNestedStack1topic209B8719858511914": {
      "Type": "AWS::SNS::Subscription",
      "Properties": {
        "Protocol": "sqs",
        "TopicArn": {
          "Fn::GetAtt": [
            "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52",
            "Outputs.nestedstackstestNestedStack1topic22C710DC4Ref"
          ]
        },
        "Endpoint": {
          "Fn::GetAtt": [
            "SubscriberQueueC193DC66",
            "Arn"
          ]
        }
      }
    },
    "NestedStack1NestedStackNestedStack1NestedStackResource7E64AD52": {
      "Type": "AWS::CloudFormation::Stack",
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region"
              },
              ".",
              {
                "Ref": "AWS::URLSuffix"
              },
              "/",
              {
                "Ref": "AssetParameters91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4S3Bucket06F505B9"
              },
              "/",
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4S3VersionKey06BEDE93"
                      }
                    ]
                  }
                ]
              },
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4S3VersionKey06BEDE93"
                      }
                    ]
                  }
                ]
              }
            ]
          ]
        },
        "Parameters": {
          "TopicNamePrefix": "Prefix1",
          "referencetonestedstackstestSubscriberQueue39409787Ref": {
            "Ref": "SubscriberQueueC193DC66"
          }
        }
      }
    },
    "NestedStack2NestedStackNestedStack2NestedStackResourceFDF82E43": {
      "Type": "AWS::CloudFormation::Stack",
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region"
              },
              ".",
              {
                "Ref": "AWS::URLSuffix"
              },
              "/",
              {
                "Ref": "AssetParameters0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22dfS3BucketB322F951"
              },
              "/",
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22dfS3VersionKeyAA9C5AF4"
                      }
                    ]
                  }
                ]
              },
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22dfS3VersionKeyAA9C5AF4"
                      }
                    ]
                  }
                ]
              }
            ]
          ]
        },
        "Parameters": {
          "TopicNamePrefix": "Prefix2"
        }
      }
    }
  },
  "Parameters": {
    "AssetParameters91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4S3Bucket06F505B9": {
      "Type": "String",
      "Description": "S3 bucket for asset \"91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4\""
    },
    "AssetParameters91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4S3VersionKey06BEDE93": {
      "Type": "String",
      "Description": "S3 key for asset version \"91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4\""
    },
    "AssetParameters91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4ArtifactHash407EE1C2": {
      "Type": "String",
      "Description": "Artifact hash for asset \"91169429f2b5b85501c7b1b9d7beeb80c9bb6f4891f4e600fcaf65a8817ce0f4\""
    },
    "AssetParameters0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22dfS3BucketB322F951": {
      "Type": "String",
      "Description": "S3 bucket for asset \"0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22df\""
    },
    "AssetParameters0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22dfS3VersionKeyAA9C5AF4": {
      "Type": "String",
      "Description": "S3 key for asset version \"0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22df\""
    },
    "AssetParameters0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22dfArtifactHash5D335705": {
      "Type": "String",
      "Description": "Artifact hash for asset \"0d0404717d8867c09534f2cf382e8e24531ff64a968afa2efd7f071ad65a22df\""
    }
  }
}